lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {DeepRecord, Key, CollectionKey, Selector, Value} from './types';
+
+type KeyValueMap = {
+    [TKey in Key | CollectionKey]: Value[TKey];
+};
+
+type BaseConnectOptions<TKey extends Key | CollectionKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, Value[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends Key | CollectionKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, Value[TKey]> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: Value[TKey] | null, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+type MergeCollection<TKey extends Key, TMap, TValue> = {
```

Comment:
Separately shouldn't it be a CollectionKey?

```ts

type MergeCollection<TKey extends CollectionKey, TMap, TValue> = {
```

Reply:
@neil-marcellini I didn't understand your question, could you elaborate more?

Reply:
Never mind, now I see that you need `TKey extends Key` to define the type such that they TKey must always be a collection key with its id set.